docs/BackupNoSQL.md

```diff
@@ -0,0 +1,153 @@
+
+## Backup operations (create the **period** for restoration)
+
+1. Check the pod status before backup.
+
+   You must check the following four points using the `kubectl get pod` command before you start the backup operation.
+   * The number of Scalar product pods (You must compare this number with the number after the backup).
+   * The Scalar product pod names in the `NAME` column (You must compare these pod names with the pod names after the backup).
+   * The Scalar product pod status is `Running` in the `STATUS` column.
+   * The restart counts of each pod in the `RESTARTS` column (You must compare these restart counts with the restart counts after the backup).
+
+1. Pause Scalar Product pods using `scalar-admin`.
+
+   Please refer to the [How to use scalar-admin on a Kubernetes environment](./BackupNoSQL.md#how-to-use-scalar-admin-on-a-kubernetes-environment) section in this document for more details on the pause operation.
+
+1. Check the pause completed time.
+
+   You must note the pause completed time to decide the concrete **period** that you can restore data using the PITR feature.
+
+1. Take a backup using the backup feature of each database.
+
+   If you enable the automatic backup and PITR feature, all you have to do is wait for about 10 seconds since the managed databases take backup automatically. This 10 seconds period is the **period** that we can restore data by PITR.
+
+1. Unpause Scalar Product pods using `scalar-admin`.
+
+   Please refer to the [How to use scalar-admin on a Kubernetes environment](./BackupNoSQL.md#how-to-use-scalar-admin-on-a-kubernetes-environment) section in this document for more details on the unpause operation.
+
+1. Check the unpause started time.
+
+   You must note the unpause started time to decide the concrete **period** that you can restore data using the PITR feature.
+
+1. Check the pod status after backup.
+
+   You must check the following four points using the `kubectl get pod` command after you finish the backup operation.
+   * The number of Scalar product pods (You must compare this number with the number before the backup).
+   * The Scalar product pod names in the `NAME` column (You must compare these pod names with the pod names before the backup).
+   * The Scalar product pod status is `Running` in the `STATUS` column.
+   * The restart counts of each pod in the `RESTARTS` column (You must compare these restart counts with the restart counts before the backup).
+
+   **If the two values are different, you must re-try the backup operation all over again.** This is because, if some pods are added or restarted, those pods run with a `unpause` state. The `unpause` state pods cause data inconsistency in the backup data.
```

Comment:
```suggestion
1. Check the following four points by running the `kubectl get pod` command before starting the backup operation:
   * **The number of ScalarDB or ScalarDL pods.** Write down the number of pods so that you can compare that number with the number of pods after performing the backup.
   * **The ScalarDB or ScalarDL pod names in the `NAME` column.** Write down the pod names so that you can compare those names with the pod names after performing the backup.
   * **The ScalarDB or ScalarDL pod status is `Running` in the `STATUS` column.** Confirm that the pods are running before proceeding with the backup. You will need to pause the pods in the next step.
   * **The restart count of each pod in the `RESTARTS` column.** Write down the restart count of each pod so that you can compare the count with the restart counts after performing the backup.
2. Pause the ScalarDB or ScalarDL pods by using `scalar-admin`. For details on how to pause the pods, see the [Details on using `scalar-admin`](./BackupNoSQL.md#details-on-using-scalar-admin) section in this guide.
3. Write down the time when the pause started. You will need to refer to that time when restoring the data by using the PITR feature.
4. Back up each database by using the backup feature. If you have enabled the automatic backup and PITR features, the managed databases will be backed up automatically in approximately 10 seconds. This 10-second period is the exact period in which you can restore data by using the PITR feature.
5. Unpause ScalarDB or ScalarDL pods by using `scalar-admin`. For details on how to unpause the pods, see the [Details on using `scalar-admin`](./BackupNoSQL.md#details-on-using-scalar-admin) section in this guide.
6. Check the unpause start time. You must check the unpause start time to confirm the exact period in which you can restore data by using the PITR feature.
7. Check the pod status after performing the backup. You must check the following four points by using the `kubectl get pod` command after the backup operation is completed.
   * **The number of ScalarDB or ScalarDL pods.** Confirm this number matches the number of pods that you wrote down before performing the backup.
   * **The ScalarDB or ScalarDL pod names in the `NAME` column.** Confirm the names match the pod names that you wrote down before performing the backup.
   * **The ScalarDB or ScalarDL pod status is `Running` in the `STATUS` column.** Confirm that the pods are running and not paused.
   * **The restart count of each pod in the `RESTARTS` column.** Confirm the counts match the restart counts that you wrote down before performing the backup
   
   **If any of the two values are different, you must retry the backup operation from the beginning.** The reason for the different values may be caused by some pods being added or restarted while performing the backup. In such case, those pods will run in the `unpause` state. Pods in the `unpause` state will cause the backup data to be transactionally inconsistent.
```

Reply:
Sorry, I don't know the correct English context, but I would like to confirm the means of `in approximately 10 seconds`.

```
the managed databases will perform back up automatically in approximately 10 seconds
```
Does this sentence mean "database will perform back up automatically **after** approximately 10 seconds" or "the automatic backup **takes** approximately 10 seconds"?
If yes, I think it is a bit different from the actual behavior.
Managed database's automatic backup feature gets database backup in the background process and we can specify the PITR point per second.
So, basically, managed databases take a backup every second.
In other words, it doesn't take 10 seconds for backup.

---

One of the purposes of this 10 seconds period is to avoid mistakes caused by clock skew between the client and the database.
Regarding ScalarDB/ScalarDL backup, we must get a backup from the period that there is no ongoing Tx because ongoing Tx causes data inconsistency.
```
                          Pause completed   Unpause started
                         /                 /
|-----------------------|-----------------|-----------------------|
 <- Ongoing Tx exists -> <- No Tx exist -> <- Ongoing Tx exists ->

 (Ongoing Tx causes data inconsistency)
```

And, we specify the PITR time based on the **Database clock**.
However, the `scalar-admin` command returns pause duration (time) based on the **Client clock**.

If the client clock and database clock are perfectly synchronized, then we don't need the long time duration.
For example, it is enough for about 2 seconds duration as a backup/restoration period as follows.
```
Client (i.e., scalar-admin) clock
+----+----+----+----+----+----+----+----+----+----+
0    1    2    3    4    5    6    7    8    9    10

Database clock
+----+----+----+----+----+----+----+----+----+----+
0    1    2    3    4    5    6    7    8    9    10

          |         |
          |         * Unpause starting at 4
          |
          * Pause started at 2
```
In this case, users can specify the time `3` as a restoration point.

However, there is no `perfectly synchronized clock` in the real world, so we need to take care of the clock skew.
For example, if the clock is off by two seconds between the client and the database as follows, there is a possibility that it causes mistakes (data inconsistency).
```
Client (i.e., scalar-admin) clock
          +----+----+----+----+----+----+----+----+----+----+
          0    1    2    3    4    5    6    7    8    9    10

Database clock
+----+----+----+----+----+----+----+----+----+----+
0    1    2    3    4    5    6    7    8    9    10

                    |         |
                    |         * Unpause starting at 4 from the client perspective
                    |
                    * Pause started at 2 from the client perspective
```
In this case, users see the pause duration that is from 2 to 4 from the output of scalar-admin based on the client clock.
And, maybe, users specify the time `3` as a restoration point in the database console.

However, in fact, the correct pause duration is from 4 to 6 in the database clock.
So, we need to specify the time `5` to specify the `No Tx exist` period.

If users specify the time `3` in the database console based on the client clock, it causes data inconsistency.
The time `3` in the database side is the duration that is the `Ongoing Tx exists`.

---

To avoid the above issue caused by the clock skew, we need to create a sufficiently long period of time (i.e., 10 seconds).

```
Client (i.e., scalar-admin) clock
          +----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+
          0    1    2    3    4    5    6    7    8    9    10   11   12   13   14   15

Database clock
+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+
0    1    2    3    4    5    6    7    8    9    10   11   12   13   14   15

               |                                                 |
               |                                                 * Unpause starting at 11 from the client perspective
               |
               * Pause started at 1 from the client perspective
```

For example, if users create a pause period as 10 seconds as the above, the pause duration is from 1 to 11 in the client clock.
And, the duration is from 3 to 13 in the database clock.
So, even if they specify the middle time `6` based on the client clock in the database console when they do the PITR, the time `6` is in the `No Tx exist` duration (3 ~ 13) in the database clock.
This doesn't cause data inconsistency.

So, the important thing is not `waiting for the automatic backup feature to complete taking a backup`.
The important thing is `make a sufficiently long period`.
This is why we wait 10 seconds.

---

Sorry for the long description so far, but I think we should update the following sentence.

```
If you have enabled the automatic backup and PITR features, the managed databases will perform back up automatically in approximately 10 seconds.
```

For example, the following.

```
If you have enabled the automatic backup and PITR features, the managed databases will perform back up automatically. And, you need to wait for approximately 10 seconds to create a sufficiently long period to avoid the clock skew issue between the client clock and the database clock.
```

What do you think?

Reply:
```
**The ScalarDB or ScalarDL pod status is `Running` in the `STATUS` column.** Confirm that the pods are running and not paused.
```

Regarding the above sentence, the pause/unpause state is Scalar products original state.
So, it cannot be observed from the Kubernetes perspective.
In other words, the `STATUS` column will not be like `PAUSED` even if we pause the Scalar product pods.
It is enough to confirm the `STATUS` is `Running`.

So, I think we can remove `Confirm that the pods are running and not paused.`.

Reply:
In this conversation, there are two comments. Please take a look at both comments.